ipsec/key_sockets.c

```diff
@@ -0,0 +1,316 @@
+/*
+ * Phoenix-RTOS --- LwIP port
+ *
+ * LwIP key sockets
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Ziemowit Leszczynski
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "key_sockets.h"
+
+#include "mbuff.h"
+#include "sadb.h"
+
+#include "lwip/opt.h"
+
+#include <errno.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/threads.h>
+
+
+#define NUM_SOCKETS            MEMP_NUM_NETCONN
+#define NUM_KEY_SOCKETS        64
+#define LWIP_KEY_SOCKET_OFFSET (LWIP_SOCKET_OFFSET + NUM_SOCKETS)
+
+
+struct key_socket_s {
+	mbuff_t *mbuff;
+	handle_t mutex;
+	semaphore_t semaphore;
+};
+typedef struct key_socket_s key_socket_t;
+
+
+static handle_t key_sockets_mutex;
+static key_socket_t *key_sockets[NUM_KEY_SOCKETS];
+
+
+static key_socket_t **tryget_key_socket(int fd)
+{
+	int i = fd - LWIP_KEY_SOCKET_OFFSET;
+
+	if (i < 0 || i >= NUM_KEY_SOCKETS)
+		return NULL;
+
+	return &key_sockets[i];
+}
+
+
+static key_socket_t **get_key_socket(int fd)
+{
+	key_socket_t **sock_ptr = tryget_key_socket(fd);
+
+	if (!sock_ptr) {
+		errno = EBADF;
+		return NULL;
+	}
+
+	return sock_ptr;
+}
+
+
+static int alloc_key_socket(key_socket_t *sock)
+{
+	unsigned int i;
+
+	for (i = 0; i < NUM_KEY_SOCKETS; ++i) {
+		mutexLock(key_sockets_mutex);
+		if (!key_sockets[i]) {
+			key_sockets[i] = sock;
+			mutexUnlock(key_sockets_mutex);
+			return i + LWIP_KEY_SOCKET_OFFSET;
+		}
+		mutexUnlock(key_sockets_mutex);
+	}
+
+	return -1;
+}
+
+
+static void free_key_socket(key_socket_t **sock_ptr)
```

Comment:
This is somewhat misleading... We pass malloced ptr to a funtion with free in name and no free() is performed inside

Reply:
Memory free is done by destroy_key_socket(), this function is freeing socket descriptor only. I will improve this code later.